datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -311,20 +311,27 @@ fn stats_cartesian_product(
     }
 }
 
-/// A stream that issues [RecordBatch]es as they arrive from the right  of the join.
+/// A stream that issues [RecordBatch]es as they arrive from the right of the join.
+/// Right column orders are preserved.
 struct CrossJoinStream {
     /// Input schema
     schema: Arc<Schema>,
-    /// future for data from left side
+    /// Future for data from left side
     left_fut: OnceFut<JoinLeftData>,
-    /// right
+    /// Right stream
     right: SendableRecordBatchStream,
-    /// Current value on the left
-    left_index: usize,
-    /// Current batch being processed from the right side
-    right_batch: Arc<parking_lot::Mutex<Option<RecordBatch>>>,
-    /// join execution metrics
+    /// Join execution metrics
     join_metrics: BuildProbeJoinMetrics,
+    /// State information
+    state: CrossJoinStreamState,
+    /// Left data
+    left_data: Vec<RecordBatch>,
```

Comment:
minor: we store both fut & data -- won't it make sense to store `stream.left_data` as enum (with initial/ready variants, wrapping future and data respectively)?